src/main/java/edu/kit/datamanager/security/filter/JwtAuthenticationToken.java

```diff
@@ -39,161 +39,167 @@
+            //as long as no additional types are added, we'll never arrive here
+            throw new InvalidAuthenticationException("JWTokens of type " + type + " are currently not supported.");
+        }
+
+        for (String claim : jwToken.getSupportedClaims()) {
+            Object value = MapUtils.getObject(claims, claim);
+            Class c = jwToken.getClassForClaim(claim);
 
-  public abstract Class getClassForClaim(String claim);
+            if (value != null && !c.isInstance(value)) {
+                throw new InvalidAuthenticationException("Claim " + claim + " is invalid. Expected type " + c);
+            }
+            jwToken.setValueFromClaim(claim, value);
+        }
 
-  public abstract void setValueFromClaim(String claim, Object value);
+        jwToken.validateToken();
 
-  public void validateToken(){
-    if(getPrincipal() == null){
-      throw new InvalidAuthenticationException("Token validatation failed. No principal assigned.");
+        jwToken.setAuthenticated(true);
+        return jwToken;
     }
 
-    validate();
-  }
+    public static List<SimpleGrantedAuthority> grantedAuthorities(Set<String> roles) {
+        if (null == roles) {
+            return new ArrayList<>();
+        }
+        return roles.stream().map(String::toString).map(SimpleGrantedAuthority::new).collect(toList());
+    }
 
-  public abstract void validate() throws InvalidAuthenticationException;
+    public abstract String[] getSupportedClaims();
 
-  public abstract TOKEN_TYPE getTokenType();
+    public abstract Class getClassForClaim(String claim);
 
-  void setPrincipalName(String principalName){
-    this.principalName = principalName;
-  }
+    public abstract void setValueFromClaim(String claim, Object value);
 
-  @Override
-  public Object getCredentials(){
-    return NOT_AVAILABLE;
-  }
+    public void validateToken() {
+        if (getPrincipal() == null) {
+            throw new InvalidAuthenticationException("Token validatation failed. No principal assigned.");
+        }
 
-  @Override
-  public final Object getPrincipal(){
-    return principalName;
-  }
+        validate();
+    }
 
-  public final String getToken(){
-    return token;
-  }
+    public abstract void validate() throws InvalidAuthenticationException;
 
-  public String getGroupId(){
-    return groupId;
-  }
+    public abstract TOKEN_TYPE getTokenType();
 
-  public void setGroupId(String groupId){
-    this.groupId = groupId;
-  }
+    void setPrincipalName(String principalName) {
+        this.principalName = principalName;
+    }
+
+    @Override
+    public Object getCredentials() {
+        return NOT_AVAILABLE;
+    }
+
+    @Override
+    public final Object getPrincipal() {
+        return principalName;
+    }
+
+    public final String getToken() {
+        return token;
+    }
 
-  @Override
-  public final void setAuthenticated(boolean authenticated){
-    super.setAuthenticated(authenticated);
-  }
+    public List<String> getGroups() {
```

Comment:
get/setGroupId is no longer available. -> This is a breaking change which should result in a new major version.